src/http.rs

```diff
@@ -382,8 +386,174 @@ pub fn build_http_trace_layer(
         .on_eos(())
 }
 
+/// Configuration of HTTP server keepalive parameters.
+#[serde_as]
+#[configurable_component]
+#[derive(Clone, Debug, PartialEq)]
+#[serde(deny_unknown_fields)]
+pub struct KeepaliveConfig {
+    /// The maximum amount of time a connection may exist before it is closed
+    /// by sending a `Connection: close` header on the HTTP response.
+    ///
+    /// A random jitter configured by `max_connection_age_jitter_factor` is added
+    /// to the specified duration to spread out connection storms.
+    ///
+    /// A value of `0` disables this feature.
+    #[serde_as(as = "serde_with::DurationSeconds<u64>")]
+    #[serde(
+        rename = "max_connection_age_secs",
+        default = "default_max_connection_age"
+    )]
+    #[configurable(metadata(docs::examples = 600))]
+    #[configurable(metadata(docs::human_name = "Maximum Connection Age"))]
+    pub max_connection_age: Duration,
+
+    /// The factor by which to jitter the `max_connection_age_secs` value.
+    ///
+    /// A value of 0.1 means that the actual duration will be between 90% and 110% of the
+    /// specified duration.
+    #[serde(default = "default_max_connection_age_jitter_factor")]
+    #[configurable(validation(range(min = 0.0, max = 1.0)))]
+    pub max_connection_age_jitter_factor: f64,
+}
+
+const fn default_max_connection_age() -> Duration {
+    Duration::from_secs(0)
+}
+
+const fn default_max_connection_age_jitter_factor() -> f64 {
+    0.1
+}
+
+impl Default for KeepaliveConfig {
+    fn default() -> Self {
+        Self {
+            max_connection_age: default_max_connection_age(),
+            max_connection_age_jitter_factor: default_max_connection_age_jitter_factor(),
+        }
+    }
+}
+
+/// A layer that limits the maximum duration of a client connection. It does so by adding a
+/// `Connection: close` header to the response if `max_connection_duration` time has elapsed
+/// since `start_reference`.
+///
+/// **Notes:**
+/// - This is intended to be used in a Hyper server (or similar) that will automatically close
+/// the connection after a response with a `Connection: close` header is sent.
+/// - This layer assumes that it is instantiated once per connection, which is true within the
+/// Hyper framework.
+
+pub struct MaxConnectionAgeLayer {
+    start_reference: Instant,
+    max_connection_age: Duration,
```

Comment:
Would it be worth storing the end reference instead of the start plus an age? i.e. store `Instant::now() + max_connection_age`

Reply:
The start reference and age are useful information to put in the debug log in my opinion.